Resource Validator fails on java:/app/ paths

# Description #
----------

When deploying a module that references an EJB via JNDI path `java:/app/moduleName/beanName`, the resource validator fails to deploy the app with exception `JNDI Lookup failed / Got null ComponentInvocation`.

Workaround is to disable resource validation with `asadmin create-jvm-options -Ddeployment.resource.validation=false`

<!--- Brief summary description of the bug or enhancement request -->

## Expected Outcome

<!-- If reporting a bug, give a detailed summary of the expected behavior the server and/or deployed applications SHOULD exhibit after executing the steps described below. If possible quote Java EE specification's sections or link to Glassfish or Payara's official documentation as evidence. -->

Deployment should pass as the JNDI names are valid and worked in versions prior to 174.

<!-- If making an enhancement request, give a detailed explanation of how this new or updated feature should work -->

## Current Outcome

````
org.glassfish.deployment.common.DeploymentException: JNDI lookup failed for the resource: Name: [java:module/env/net.dudits.payara.issue688.EjbProducer/firstEjb], Lookup: [java:app/ejbmodule/FirstEjb], Type: [Session]
	at com.sun.enterprise.deployment.util.ResourceValidator.validateJNDIRefs(ResourceValidator.java:874)
	at com.sun.enterprise.deployment.util.ResourceValidator.validateResources(ResourceValidator.java:736)
	at com.sun.enterprise.deployment.util.ResourceValidator.event(ResourceValidator.java:141)
	at org.glassfish.kernel.event.EventsImpl.send(EventsImpl.java:131)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:424)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:220)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:508)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:544)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:540)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:539)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:570)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:562)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:561)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1469)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:111)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1851)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1727)
	at org.glassfish.admin.rest.resources.admin.CommandResource.executeCommand(CommandResource.java:407)
	at org.glassfish.admin.rest.resources.admin.CommandResource.execCommandSimpInMultOut(CommandResource.java:234)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:384)
	at org.glassfish.admin.rest.adapter.JerseyContainerCommandService$3.service(JerseyContainerCommandService.java:173)
	at org.glassfish.admin.rest.adapter.RestAdapter.service(RestAdapter.java:179)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:483)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:180)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:539)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.naming.NamingException: Lookup failed for 'java:app/ejbmodule/FirstEjb' in SerialContext[myEnv={java.naming.factory.initial=com.sun.enterprise.naming.impl.SerialInitContextFactory, java.naming.factory.state=com.sun.corba.ee.impl.presentation.rmi.JNDIStateFactoryImpl, java.naming.factory.url.pkgs=com.sun.enterprise.naming}
	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:496)
	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:442)
	at javax.naming.InitialContext.lookup(InitialContext.java:417)
	at javax.naming.InitialContext.lookup(InitialContext.java:417)
	at com.sun.enterprise.deployment.util.ResourceValidator.validateJNDIRefs(ResourceValidator.java:869)
	... 65 common frames omitted
Caused by: javax.naming.NamingException: Invocation exception: Got null ComponentInvocation 
	at com.sun.enterprise.naming.impl.GlassfishNamingManagerImpl.getComponentId(GlassfishNamingManagerImpl.java:862)
	at com.sun.enterprise.naming.impl.GlassfishNamingManagerImpl.lookup(GlassfishNamingManagerImpl.java:734)
	at com.sun.enterprise.naming.impl.JavaURLContext.lookup(JavaURLContext.java:167)
	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:476)
	... 69 common frames omitted
````
<!-- If reporting a bug, give a detailed summary of the actual behavior the server and/or deployed applications exhibit after executing the steps described below. Please put emphasis on any unwanted results. -->

<!-- If making an enhancement request, explain the drawbacks and disadvantages of the targeted feature (or lack of it) -->

## Steps to reproduce (Only for bug reports) 

My reproducer for older issue can be reused. 

1. Build project in my testcases repo: https://github.com/pdudits/payara-testcases/tree/master/issue-668
2. asadmin deploy ejbmodule/target/ejbmodule.jar
3. asadmin deploy webmodule/target/webmodule.war

`FirstEjb` is an ejb with Appendable as business interface. It's global jndi name is therefore `java:global/ejbmodule/FirstEjb`.

`EjbProducer` in `webmodule` injects this EJB via `@EJB(lookup = "java:app/ejbmodule/FirstEjb")`. Since this is not an ear, `java:app` is the same as `java:global`.

<!-- 

Describe the test to reproduce the bug in a series of steps. Make each step simple to follow by describing configuration changes, commands to run or simple instructions; for example:

1 -**  Start the domain

    ./asadmin start-domain payaradomain

2 -**  Configure an HTTP network listener in the admin console:

[Attach screenshots of Payara's Server console, to illustrate]

3 -**  Make changes to the domain.xml configuration:

	<java-config classpath-suffix="" debug-options="-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=9009" system-classpath="">
        <jvm-options>-XX:MaxPermSize=512m</jvm-options>
        <jvm-options>-server</jvm-options>
		...
	</java-config>

-->

### Samples

https://github.com/pdudits/payara-testcases/tree/master/issue-668

<!-- Include a link to a [SCCE](http://sscce.org/ "Short, Self-Contained, Correct Example") that helps reproduce the issue faster. Structuring a Maven project is strongly recommended if possible -->

## Context (Optional)

<!-- Give details on this issue has affected you, for example: What are you trying to accomplish?
Providing context helps us come up with a solution that is most useful for your scenario. -->

## Environment ##

- **Payara Version**: 4.1.2.181
- **Edition**: Full<!-- Full/Web/Blue/Micro -->
- **JDK Version**: 8u131<!-- 6/7/8 uXX - Oracle/IBM/OpenJDK -->
- **Operating System**: Windows<!-- Windows / Linux / Mac -->
- **Database**: N/A<!-- Oracle/MySQL/MariaDB/PostGres/DB2/SQL Server -->

--- a/appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java
+++ b/appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java
@@ -66,6 +66,8 @@ import java.util.*;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import org.glassfish.hk2.runlevel.RunLevel;
+import org.glassfish.internal.api.JavaEEContextUtil;
+import org.glassfish.internal.api.JavaEEContextUtil.Context;
 import org.glassfish.internal.api.PostStartupRunLevel;
 
 /**
@@ -110,6 +112,9 @@ public class ResourceValidator implements EventListener, ResourceValidatorVisito
 
     @Inject
     private Domain domain;
+    
+    @Inject
+    JavaEEContextUtil contextUtil;
 
     private static LocalStringManagerImpl localStrings = new LocalStringManagerImpl(ResourceValidator.class);
 
@@ -138,7 +143,21 @@ public class ResourceValidator implements EventListener, ResourceValidatorVisito
             AppResources appResources = new AppResources();
             //Puts all resouces found in the application via annotation or xml into appResources
             parseResources(appResources);
-            validateResources(appResources);
+            
+            // Ensure we have a valid component invocation before triggering lookups
+            String componentId = null;
+            for (BundleDescriptor bundleDescriptor : application.getBundleDescriptors()) {
+                if (bundleDescriptor instanceof JndiNameEnvironment) {
+                    componentId = DOLUtils.getComponentEnvId((JndiNameEnvironment) bundleDescriptor);
+                    if (componentId != null) {
+                        break;
+                    }
+                }
+            }
+            contextUtil.setInstanceComponentId(componentId);
+            try (Context ctx = contextUtil.pushContext()) {
+                validateResources(appResources);
+            }
         }
     }
 
@@ -841,12 +860,28 @@ public class ResourceValidator implements EventListener, ResourceValidatorVisito
         if (isResourceInDomainXML(jndiName) || isDefaultResource(jndiName)) {
             return;
         }
+       
 
         // Managed Bean & EJB portable JNDI names
         if (jndiName.startsWith(ResourceConstants.JAVA_MODULE_SCOPE_PREFIX) || jndiName.startsWith(ResourceConstants.JAVA_APP_SCOPE_PREFIX)) {
+            // first look with the name
+            if (namespace.find(jndiName, env)) {
+                return;
+            }
+            
             String newName = convertModuleOrAppJNDIName(jndiName, resource.getEnv());
             if (namespace.find(newName, env)) {
                 return;
+            } else {
+                // try actual lookup
+                try {
+                    InitialContext ctx = new InitialContext();
+                    ctx.lookup(newName);
+                    return;
+                } catch(NamingException ne) {
+                    
+                }
+                
             }
         }
 

